flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java

```diff
@@ -87,9 +87,14 @@ protected <K> CheckpointableKeyedStateBackend<K> restore(
         String subtaskName = env.getTaskInfo().getTaskNameWithSubtasks();
         ExecutionConfig executionConfig = env.getExecutionConfig();
 
+        env.getAsyncOperationsThreadPool();
+
         ChangelogStateFactory changelogStateFactory = new ChangelogStateFactory();
         CheckpointableKeyedStateBackend<K> keyedStateBackend =
                 ChangelogBackendRestoreOperation.restore(
+                        env.getJobID(),
+                        env.getAsyncOperationsThreadPool(),
+                        env.getTaskManagerInfo().getConfiguration(),
```

Comment:
I debug this locally, there is only  `ENABLE_CHANGE_LOG_FOR_APPLICATION` in `JobConfiguration`, and the `PERIODIC_MATERIALIZATION_INTERVAL`  that `ChangelogHandleReaderWithCache` needs is not in `TaskManagerInfo` too. 

I think this should be `env.getExecutionConfig().getPeriodicMaterializeIntervalMillis()`.

Reply:
Thanks for checking this @fredia .

Do you mean changing `StateChangelogStorageFactory` interface and passing `ExecutionConfig` to `createStorageView` instead of `Configuration`?

That would require any new configuration parameter to be placed in `ExecutionConfig`. 
That would be problematic especially for non-bundled `StateChangelogStorageFactory` implementations.

After, #20160 (FLINK-28286), it should be env.getJobConfiguration()` ideally merged with `env.getTaskManagerInfo().getConfiguration()`, right?

If so, I think this merging can either be implemented in this PR or in FLINK-26372.

Reply:
> Do you mean changing StateChangelogStorageFactory interface and passing ExecutionConfig to createStorageView instead of Configuration?

I tend to put `PERIODIC_MATERIALIZATION_INTERVAL` directly as a parameter instead of `xxConfiguration`.

> it should be env.getJobConfiguration()ideally merged withenv.getTaskManagerInfo().getConfiguration()`, right?

I'm not sure.  Currently, most configuration is in `env.getTaskManagerInfo().getConfiguration()` or `env.getTaskConfiguration()`,  which is better to merge?

Reply:
Hi @fredia , thanks for reply. I'm not suggest pass PERIODIC_MATERIALIZATION_INTERVAL directly. StateChangelogStorage may have different implementations, each one has different  options. I think an implementation-specific configuration should not be exposed in the interface.